test/system/src/test/java/io/pravega/test/system/AbstractFailoverTests.java

```diff
@@ -97,22 +95,90 @@
         final CompletableFuture<Void> newWritersComplete = new CompletableFuture<>();
         final CompletableFuture<Void> readersComplete = new CompletableFuture<>();
         final List<CompletableFuture<Void>> txnStatusFutureList = synchronizedList(new ArrayList<>());
-        final AtomicBoolean txnWrite = new AtomicBoolean(false);
+
+        final ConcurrentSet<UUID> committingTxn = new ConcurrentSet<>();
+        final ConcurrentSet<UUID> abortedTxn = new ConcurrentSet<>();
+        final AtomicLong eventData = new AtomicLong();
+        final boolean txnWrite;
+
+        private final ConcurrentHashMap<Long, Integer> eventMap = new ConcurrentHashMap<>();
+
+        TestState(boolean txnWrite) {
+            this.txnWrite = txnWrite;
+        }
+
+        void eventWritten(Long event) {
+            eventMap.putIfAbsent(event, 0);
+        }
+
+        void eventRead(Long event) {
+            eventMap.compute(event, (x, y) -> {
+                if (y == null) {
+                    return 1;
+                } else {
+                    return y + 1;
+                }
+            });
+        }
+
+        Long getNext() {
+            return eventData.getAndIncrement();
+        }
+
+        int getEventWrittenCount() {
+            return eventMap.size();
+        }
+
+        int getEventReadCount() {
+            return eventMap.values().stream().mapToInt(Integer::intValue).sum();
+        }
+
+        void printAnomalies() {
```

Comment:
The idea of having a method that centralizes the verification that nothing has gone wrong is great. I wonder if we should strengthen it, though. We could call it `checkForAnomalies` and fail the test in the case anything fails. Currently, we are just logging.

Reply:
done